### DISCUSSION TITLE
Discussion for application to include and compressed

### DISCUSSION THREAD

**Discussion by tyama501** [jbruchon/elks#1870] [General]
Discussion relating to

CONFIG_APPS_xxx
https://github.com/ghaerr/elks/pull/1853

and 

CONFIG_APPS_COMPRESS
https://github.com/ghaerr/elks/pull/1867

**Comment by ghaerr** [comment]
If you like, I can implement two new tags to be used in elkscmd/Applications (`:1232k` and `:1232c`) which would allow you to experiment with adding files two the PC-98 distribution using CONFIG_APPS_BY_IMAGSZ and CONFIG_APPS_COMPRESS  rather than CONFIG_APPS_xxx.

In general, it would be good if the `:1200k` tag in elkscmd/Applications would be used to build the base disk image, and then :1232k was added for PC-98 uncompressed and :1232c for compressed image.

Ultimately, we just need to know the additional files that need to be added to the 1232k distribution when CONFIG_APPS_BY_IMAGESZ is used, rather than the current CONFIG_APPS_xxx settings. That will allow the images to be identical, which is the goal - that the newer select-by-size option can be used with identical results.

If there is too much difference between the IBM PC :1200k image and the desired PC-98 1232k image (or, if the PC-98 1440k image is desired to be different from the IBM PC 1440k image), then I would propose a completely custom solution for selecting applications for PC-98 by using a new elkscnd/Applications.pc98 file when CONFIG_ARCH_PC98 is set.

What do you think?

**Reply by tyama501** [reply]
Yes, I started trying CONFIG_APPS_BY_IMAGESZ now. I agree to use it.
I can try this now assuming 1200k application is included, right?
I will post the difference with and without compression.

I might add Japanese applications in near future for 1232k and 1440k so I think it would be nice to have
elkscmd/Applications.pc98 file when CONFIG_ARCH_PC98 is set.
(Maybe 1232k and 1232c tag are needed even for that case)

**Reply by ghaerr** [reply]
> I can try this now assuming 1200k application is included, right?

Actually, in the first PR, only 360k and 1440k disks will produce anything different with compression. So I will implement today a PR adding :1200k, :1200c and test that, which will then add more files to the 1200k compressed floppy. I will then implement :1232k and :1232c which will allow you to experiment with adding PC-98 specific files to 1232k disk.

> I might add Japanese applications in near future for 1232k and 1440k so I think it would be nice to have
elkscmd/Applications.pc98 file when CONFIG_ARCH_PC98 is set.

Lets wait on that until we get the normal 1232k PC-98 distribution working. The issue with adding japanese-specific applications outside of the ELKS elkscmd/ build is more complicated, as they aren't built automatically with ELKS. So that will have to be handled as an extra step. IMO, the best way to handle that might be just to leave extra space on the image so that you can run an additional pass that adds your custom applications from your repos after having ELKS create the initial 1232k disk.

**Reply by tyama501** [reply]
Well, it seemed compressed nicely... 

CONFIG_APPS_1232K Without compression
<img width="323" alt="fd1232_without_compress" src="https://github.com/ghaerr/elks/assets/61556504/6b30b8fa-75db-4010-8725-c6298ae8dc6b">

CONFIG_APPS_1232K With compression
![fd1232_with_compress](https://github.com/ghaerr/elks/assets/61556504/1e8a3cdf-5944-4326-8ca5-1568574a2361)

**Reply by tyama501** [reply]
Could you add CONFIG_APPS_1232K in make menuconfig?
I am now modifying autoconfig directly.

**Reply by tyama501** [reply]
It seems that ttyclock won't work for PC-98.
![image](https://github.com/ghaerr/elks/assets/61556504/a21ff40f-9a38-4bab-91ce-460ac6c08838)

I will do speed comparison on real hardware with sl.

**Reply by ghaerr** [reply]
Does ttypong or ttytetris display at all on PC-98?

> Could you add CONFIG_APPS_1232K in make menuconfig?

Yes, I'll add that along with the 1200k and 1232k tags

**Reply by ghaerr** [reply]
#1871 Adds CONFIG_APPS_1232K as well as the additional tags discussed.

**Reply by tyama501** [reply]
Thank you!

I think attron and others in cursors need to be rewritten for PC-98 to work these tui applications.

I would like to remove these and some network application that does not work for PC-98 and add nx application instead for now.

So should I add ifdef in the Application list?
or do we going to have separate list?

**Reply by ghaerr** [reply]
>  think attron and others in cursors need to be rewritten for PC-98 to work these tui applications.

I thought that the PC-98 console emulated the full set of ANSI escape sequences. Is that not the case? I see that console-direct-pc98.c does `#include "console.c"` so it should work, with the possible exception of enable/disable cursor, which isn't a big deal.

> I would like to remove these and some network application that does not work for PC-98 and add nx application instead for now.
> So should I add ifdef in the Application list? or do we going to have separate list?

Studying the elkscmd/Applications file and looking at the following code in elkscmd/Make.install:
```
ifdef CONFIG_APPS_1232K         
    TAGS = :boot|:360k|:net|:720k|:1200k|:1232k|
ifdef CONFIG_APPS_COMPRESS
    TAGS += :360c|:1232c|:1440k|
endif
endif
```
I think probably rather than maintaining a separate list (and ifdefs won't work), it will be better to keep a single elkscmd/Applications file and use `:1232k` in every line (just after `:1200k` in most cases) for every program you would like installed on PC-98 1232k image. This will allow you to keep exactly what you want, discard the network utilities, and add nx applications.

After you add :1232k, your default non-compressed distribution should work, by editing the above elkscmd/Make.install to something like:
```
ifdef CONFIG_APPS_1232K         
    TAGS = :boot|:360k|:720k|:1232k|
ifdef CONFIG_APPS_COMPRESS
    TAGS += :360c|:1232c|
endif
endif
```
After getting the non-compressed distribution to work, then go ahead and add files using `:1232c` that will fit only on the compressed distribution.

Try that and post a PR with the changes, we can then discuss and commit it.

**Reply by tyama501** [reply]
The ttyclock uses wattron and wbkgdset to draw number.

In curses2.c, these uses bg/A_BLINK/A_BOLD.

PC-98 does not have bg.
The conv_attr in console-direct-pc98.c does some conversion but may be that is not enough.
The tetris does not show anything. I assume similar reason.

I will try tags.
Thank you.

**Reply by ghaerr** [reply]
> In curses2.c, these uses bg/A_BLINK/A_BOLD.

A_BLINK and A_BOLD do nothing in the current curses, other than possibly turning off all attributes (with `ESC [ m`).

> The ttyclock uses wattron and wbkgdset to draw number.

`wbkgdset` uses reverse video (`ESC [ 7 m`) and `wattron` set FG color (which is previously reversed by `wbkgdset`) using the ANSI color sequences `ESC [  fg; bg m` sequence.

> PC-98 does not have bg.

You mean that one can't set a background color? Or that there is no reverse video? Perhaps this can be simulated at the low level by reversing the fg and bg color. The console.c code will be parsing the ANSI sequences just like IBM PC, so I assume you're saying that setting the reverse video attribute isn't working (yet) on PC-98 hardware.

> The tetris does not show anything. I assume similar reason.

Both ttytetris and ttyclock use reverse video ESC 7 m followed by ANSI color in order to get "block character + fg color", so yes.

**Reply by tyama501** [reply]
I will look conv_attr when I have more time since I forgot how I implemented reverse character but at least there is no bg on PC -98 because the text console is overlapped on the graphic display and the bg is the 
graphic display.

**Reply by ghaerr** [reply]
Hello @tyama501,

> I would like to remove these and some network application that does not work for PC-98 and add nx application instead for now.

I think I have found a way we can modify elkscmd/Applications to remove just `:net` applications and allow you to add nx applications and others for PC-98, without having to go through and add `:1232k` to almost every `:1200k` entry. I will post that so that you don't have to go through the process.

Thank you!

**Reply by tyama501** [reply]
I would also like to add new beep which is not included in 1200K :)
Thank you.

**Reply by tyama501** [reply]
Current fd1232 compressed image is 185KB free.

![image](https://github.com/ghaerr/elks/assets/61556504/442fb9aa-af93-4dc2-9b6c-5c1e53608f37)

Maybe I would like to free 100KB more to add Japanese applications and fonts later.
One of the candidates of remove is nxclock that cannot show local time.
I will think about it a little more.

Thank you.

**Reply by ghaerr** [reply]
One idea to add more space to the compressed image would be to remove the automatic addition of all the `:1440k` files and instead only add those you want, using `:1232c`. This would be done by removing `:1440k|` in the following line in elkscmd/Make.install:
```
    TAGS += :360c|:1232c|:1440k|
```
Feel free to remove `nxclock` from the 1232k image. We might also move things like `tee`, `cal` and `head` from :720k to :1440k to help free some space.

**Reply by ghaerr** [reply]
Another idea would be to use `elks-compress` on the .FNT file and then decompress it in-place in far memory. This is a bit more complicated, but essentially would work by reading in the compressed .FNT file as you normally do into an area that is just 16 bytes larger than needed. The `decompress` function in elks/fs/exodecr.c would then be called to decompress the data before being used. It is decompressed in the same buffer that it resides in, pretty cool. More examples on how that is done is in the elkscmd/sys_utils/decomp.c source if you are interested.

**Comment by tyama501** [comment]
I tested with PC-9801RX21 with intel 286, 12MHz.

Here are video.
sl witout compress 
https://github.com/ghaerr/elks/assets/61556504/2ac7b133-e8e7-4a21-a7c0-e9ae0f31cb4a

![sl_without_compress](https://github.com/ghaerr/elks/assets/61556504/f4b447c4-93b1-4110-ac01-fe03536d15be)


sl with compress
https://github.com/ghaerr/elks/assets/61556504/bb667489-1947-4ad9-8205-c0d9a2491af7

![sl_with_compress](https://github.com/ghaerr/elks/assets/61556504/acf613e7-dc82-4144-b4e3-7e9c99e9e328)

Only 2.3 second difference.
Not bad at all.

Thank you!
